### PR TITLE
Add possibility to input multiple sentences

### PR DESCRIPTION
**Pull request by dldk-gael**
In order to be able to fully benefit from parallelization, I propose to add the possibility to input the sentences into the transformer models by batch. 

The work is not finished yet and I still have to pass the CI test (I will proceed as you suggest in #3) but it is working and would like to have your opinion before going further. 

In order to make minimal change to your code, for now tokens_score return a list of log_probs, ids, tokens (one item for each sentence). However, it would be much more efficient to make the reduction when we still have the log_probs scores as a tensor. One possibility would be that _tokens_log_prob now return a tensor of shape (number sentences, max sentences length). What do you think of that ?

Also, I did not use the pad_sequence function from torch.nn.utils.rnn but recode it in order that it returns a mask (with 1 at the place of the padding value) which is useful latter in the code to remove the score from padding value. I think this function should not be in GPT2LMScorer class but somewhere else.

### CODE REVIEW

File: lm_scorer/models/abc/base.py [hunk 1]
```diff
@@ -11,31 +11,38 @@ def __init__(self, model_name: str, **kwargs: Any) -> None:
...
-            score = log_probs.logsumexp(0) - math.log(tlen)
-        elif reduce == "gmean":
-            score = log_probs.mean(0)
-        elif reduce == "hmean":
-            score = log_probs.neg().logsumexp(0).neg() + math.log(tlen)
-        else:
-            raise ValueError("Unrecognized scoring strategy: %s" % reduce)
+        self, sentences: Union[str, Iterable[str]], log: bool = False, reduce: str = "prod"
+    ) -> Union[float, List[float]]:
+        sentences = [sentences] if type(sentences) == str else sentences
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:16]:
```suggestion
        sentences = [text] if type(text) == str else text
```

File: lm_scorer/models/abc/base.py [hunk 2]
```diff
@@ -11,31 +11,38 @@ def __init__(self, model_name: str, **kwargs: Any) -> None:
...
+            else:
+                raise ValueError("Unrecognized scoring strategy: %s" % reduce)
+
+            if not log:
+                score = score.exp()
+
+            scores.append(score.item())
 
-        return score.item()
+        return scores[0] if len(scores) == 1 else scores
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:40]:
```suggestion
        return scores[0] if type(text) == str else scores
```
Otherwise if the user inputs a list of strings with a single string then you return a string instead of a list of string. We want to give in output the same type as we receive in input.

File: lm_scorer/models/abc/base.py [hunk 3]
```diff
@@ -11,31 +11,38 @@ def __init__(self, model_name: str, **kwargs: Any) -> None:
...
 
     def tokens_score(
         self, text: str, log: bool = False
     ) -> Tuple[List[float], List[int], List[str]]:
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:44]: I would like this function to return depending on wether text is a list of string or a string like you did for the sentence_score function
```python
Union[
  Tuple[List[float], List[int], List[str]],
  List[Tuple[List[float], List[int], List[str]]]
]:
```

**Comment by dldk-gael** [comment]
I have taking into account all your comments. There is still one typing issue, however not sure how to resolve this one. Am i supposed to use a # type ignore or is there an other possibility ?

File: lm_scorer/models/abc/base.py [hunk 4]
```diff
@@ -10,38 +10,76 @@ class LMScorer(ABC):
...
+        List[Tuple[List[float], List[int], List[str]]],
+    ]:
+        sentences = [text] if isinstance(text, str) else text
+        outputs = [
+            (
+                log_probs.exp().tolist() if not log else log_probs.tolist(),
+                ids.tolist(),
+                tokens,
+            )
+            for log_probs, ids, tokens in self._tokens_log_prob(sentences)
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:79]: Please rewrite this with the for loop outside and use `outputs.append`.

```python
outputs = []
for scores, ids, tokens in self._tokens_log_prob(sentences):
  if not log:
    scores = scores.exp()
  outputs.append(scores.tolist(), ids.tolist(), tokens)
```

File: lm_scorer/models/abc/base.py [hunk 5]
```diff
@@ -10,38 +10,76 @@ class LMScorer(ABC):
...
+        ]
 
-        return scores.tolist(), ids.tolist(), tokens
+        return outputs[0] if isinstance(text, str) else text
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:82]:
```suggestion
        return outputs[0] if isinstance(text, str) else outputs
```

**Comment by simonepri** [comment]
> Am i supposed to use a # type ignore or is there an other possibility ?

No, it is correctly pointing out a bug in the code

**Comment by dldk-gael** [comment]
Will add some more tests concerning this new batch features in test_gpt2.

File: lm_scorer/models/gpt2.py [hunk 6]
```diff
@@ -18,46 +19,89 @@ def _build(self, model_name: str, options: Dict[str, Any]) -> None:
...
+            for i in range(len(sentences))
+        ]  # type: ignore
 
-    # @overrides
+    def _tokens_log_prob(
+        self, sentences: List[str]
+    ) -> List[Tuple[torch.FloatTensor, torch.LongTensor, List[str]]]:
+
+        output = []
+        for i in range(len(sentences) // self.batch_size):
```
Comment (simonepri) [inline lm_scorer/models/gpt2.py:94]: Use this instead
```python
for i in range(0, len(sentences), self.batch_size):
  batch = sentences[i:i+self.batch_size]
```

File: lm_scorer/models/abc/base.py [hunk 7]
```diff
@@ -70,16 +70,13 @@ def tokens_score(
...
-                log_probs.exp().tolist() if not log else log_probs.tolist(),
-                ids.tolist(),
-                tokens,
-            )
-            for log_probs, ids, tokens in self._tokens_log_prob(sentences)
-        ]
-
-        return outputs[0] if isinstance(text, str) else text
+        outputs = []
+        for scores, ids, tokens in self._tokens_log_prob(sentences):
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:74]:
```suggestion
        for log_probs, ids, tokens in self._tokens_log_prob(sentences):
            log_probs = scores  # type: torch.Tensor # type: ignore
```

File: lm_scorer/models/abc/base.py [hunk 8]
```diff
@@ -70,16 +70,13 @@ def tokens_score(
...
+            if not log:
+                scores = scores.exp()  # type: torch.Tensor # type: ignore
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:76]:
```suggestion
                scores = scores.exp()
```

**Comment by simonepri** [comment]
Would you mind if I ask you to split this PR in two?

It would be convinient to have a first PR (we can use this one) with the API changes + GPT2 implemented as just a for loop on the old single sentence code.

Then the second PR will actually optimize GPT2 using batching.
So we have the new API merged for #3

File: lm_scorer/models/gpt2.py [hunk 9]
```diff
@@ -57,6 +57,13 @@ def _tokens_log_prob(
 
         return log_probs[0], ids[0], tokens  # type: ignore
 
+    def _tokens_log_prob(
+        self, sentences: List[str]
+    ) -> List[Tuple[torch.FloatTensor, torch.LongTensor, List[str]]]:
+        return [
```
Comment (simonepri) [inline lm_scorer/models/gpt2.py:63]:
```python
return list(map(self._tokens_log_prob_single_sentence, text))
```

File: lm_scorer/models/abc/base.py [hunk 10]
```diff
@@ -10,38 +10,79 @@ class LMScorer(ABC):
...
+        self, text: List[str], log: bool = False, reduce: str = "prod"
+    ) -> List[float]:
+        ...
+
+    def sentence_score(
+        self, text: Union[str, List[str]], log: bool = False, reduce: str = "prod",
+    ) -> Union[float, List[float]]:
+        sentences = [text] if isinstance(text, str) else text
+        if len(sentences) == 0:
+            raise ValueError("Try to score a empty list of sentences")
```
Comment (simonepri) [inline lm_scorer/models/abc/base.py:30]:
```suggestion
            return []
```

File: tests/models/test_gpt2.py [hunk 11]
```diff
@@ -51,6 +51,21 @@ def should_work_on_an_empty_sentence():
         score = scorer.sentence_score("", reduce="hmean", log=True)
         assert score <= 0.0
 
+    def should_work_on_an_empty_list():
+        scorer.sentence_score([])
```
Comment (simonepri) [inline tests/models/test_gpt2.py:55]:
```suggestion
        assert scorer.sentence_score([]) == []
```

File: tests/models/test_gpt2.py [hunk 12]
```diff
@@ -70,6 +85,9 @@ def should_work_on_an_empty_sentence():
         assert len(tokens) == 1, tokens
         assert scores[0] <= 0.0
 
+    def should_work_on_an_empty_list():
+        scorer.tokens_score([])
```
Comment (simonepri) [inline tests/models/test_gpt2.py:89]:
```suggestion
        assert scorer.tokens_score([]) == []
```

**Comment by simonepri** [comment]
@dldk-gael Thanks!

**Comment by dldk-gael** [comment]
No problem and thank you for your patience, I was not familiar with all those tools and good practices, I learned a lot from your code.
